src/snc/agents/activity_rate_to_mpc_actions/fox_mpc.py

```diff
@@ -0,0 +1,178 @@
+        num_activities_per_resource = []  # type: List[int]
+
+        for c in physical_constituency_matrix:
+            activities_c = np.nonzero(c)[0]
+            assert activities_c.size > 0
+
+            activities_per_resource += [activities_c.tolist()]
+            num_activities_per_resource += [activities_c.size]
+
+        return activities_per_resource, num_activities_per_resource
+
+    def obtain_actions(self, **kwargs) -> types.ActionProcess:
+        """
+        This method implements the abstract method from super class `ActionMPCPolicy`.
+        It first gathers the feedback information namely the number of times each activity has to
+        be performed (i.e. 'sum_actions') and the current state ('state'). Then, it calls its own
+        method to return a single action vector.
+
+        :return: actions: binary indicator matrix with number of rows equal to the number of
+            non-idling activities (i.e. num of columns of the constituency matrix), and number of
+            columns equal to number of time steps to perform MPC.
+        """
+        assert 'state' in kwargs, "Current state ('state') is required to be passed as parameter."
+
+        state = kwargs["state"]
+        x_eff = kwargs["x_eff"]
+        x_star = kwargs["x_star"]
+        print(state.ravel().astype(int),x_eff.ravel().astype(int),x_star.ravel().astype(int))
+        idling_set = kwargs["r_idling_set"]
+        draining_resources = kwargs["draining_resources"]
+        print(idling_set)
+        print(self.activities_per_resource)
+        x_target = x_star if len(idling_set) > 0 else x_eff
+        buffer_weights = list(np.maximum(x_target - state, 0).astype(int).ravel())
+        print(buffer_weights)
+        actions = np.zeros((self.n_activities,1))
+
+        for r in draining_resources:
+            actions_list = self._get_actions_list(r,state,buffer_weights,False)
+            random.shuffle(actions_list)
+            decided_action,_ = actions_list[0]
+            actions[decided_action,0] = 1
+
+
+        for r in range(len(self.activities_per_resource)):
+            if r in draining_resources:
+                continue
+            actions_list = self._get_actions_list(r,state,buffer_weights, r in idling_set)
+            if not actions_list:
+                continue
+            random.shuffle(actions_list)
+            print(r,actions_list)
+            decided_action,_ = actions_list[0]
+
+
+            actions[decided_action,0] = 1
+        print()
+        return actions
+
+    def _get_actions_list(self,r,state,buffer_weights,r_in_idling_set):
+
+        actions_list = []
+        max_weight = 0
+
+        has_starved_activities = False
+        for a in self.activities_per_resource[r]:
+            if a in self.exit_activities:
+                continue
+            source_buffer = self.activities_to_source_buffers[a]
+            target_buffer = self.activities_to_target_buffers[a]
```

Comment:
Are you taking into account that there could be multiple source and target buffers here?